src/dbt_client/dbtIntegration.ts

```diff
@@ -305,52 +346,99 @@ export class DBTCommandExecutionInfrastructure {
     } catch (_) {}
   }
 
-  async addCommandToQueue(command: DBTCommand) {
-    this.queue.push({
+  createQueue(queueName: string) {
+    this.queues.set(queueName, []);
+  }
+
+  async addCommandToQueue(queueName: string, command: DBTCommand) {
+    this.queues.get(queueName)!.push({
       command: async (token) => {
         await command.execute(token);
       },
       statusMessage: command.statusMessage,
       focus: command.focus,
+      token: command.token,
+      showProgress: command.showProgress,
     });
-    this.pickCommandToRun();
+    this.pickCommandToRun(queueName);
   }
 
-  private async pickCommandToRun(): Promise<void> {
-    if (!this.running && this.queue.length > 0) {
-      this.running = true;
-      const { command, statusMessage, focus } = this.queue.shift()!;
-
-      await window.withProgress(
-        {
-          location: focus
-            ? ProgressLocation.Notification
-            : ProgressLocation.Window,
-          cancellable: true,
-          title: statusMessage,
-        },
-        async (_, token) => {
-          try {
-            await command(token);
-          } catch (error) {
-            window.showErrorMessage(
-              extendErrorWithSupportLinks(
-                `Could not run command '${statusMessage}': ` +
-                  (error as Error).message +
-                  ".",
-              ),
-            );
-            this.telemetry.sendTelemetryError("queueRunCommandError", error, {
-              command: statusMessage,
-            });
-          }
-        },
-      );
-
-      this.running = false;
-      this.pickCommandToRun();
+  private async pickCommandToRun(queueName: string): Promise<void> {
+    const queue = this.queues.get(queueName)!;
+    const running = this.queueStates.get(queueName);
+    if (!running && queue.length > 0) {
+      this.queueStates.set(queueName, true);
+      const { command, statusMessage, focus, showProgress } = queue.shift()!;
+      const commandExecution = async (token?: CancellationToken) => {
+        try {
+          await command(token);
+        } catch (error) {
+          window.showErrorMessage(
+            extendErrorWithSupportLinks(
+              `Could not run command '${statusMessage}': ` + error + ".",
+            ),
+          );
+          this.telemetry.sendTelemetryError("queueRunCommandError", error, {
+            command: statusMessage,
+          });
+        }
+      };
+
+      if (showProgress) {
+        await window.withProgress(
+          {
+            location: focus
+              ? ProgressLocation.Notification
+              : ProgressLocation.Window,
+            cancellable: true,
+            title: statusMessage,
+          },
+          async (_, token) => {
+            await commandExecution(token);
+          },
+        );
+      } else {
+        await commandExecution();
+      }
+      this.queueStates.set(queueName, false);
+      this.pickCommandToRun(queueName);
```

Comment:
not sure about this. This might hit max recursion depth limit. Maybe better to use while loop.

Reply:
i guess should be fine if queues are small.

Reply:
This is not really new code. we have been using it in core for ages. being able to support more than one queue is new.